### PR TITLE
finished R type and asembly testing all of them 

### PR DESCRIPTION
**Pull request by benziongoldstein**
# RISC-V R-type Operations Implementation

## Overview
Implemented and tested all RISC-V R-type operations in the decoder and ALU. Each operation was verified with assembly test programs to ensure correct functionality.

## Implemented Operations
- **Arithmetic**: ADD, SUB
- **Shifts**: SLL, SRL, SRA
- **Comparisons**: SLT, SLTU
- **Bitwise**: AND, OR, XOR

## Testing Methodology
- Created assembly test programs for each operation
- Used the builder script (`python3 build/builder.py core -all-asm`) to compile and simulate
- Verified results through register updates and waveform analysis
- Tested edge cases (negative numbers, zero, all ones) where applicable

## Key Features
- All operations follow RISC-V specification
- Proper handling of signed/unsigned operations
- Correct bit manipulation for shift operations
- Accurate arithmetic and logical operations

## Verification
Each operation was verified with multiple test cases:
- Basic functionality
- Edge cases
- Special cases (e.g., SRA vs SRL for negative numbers)
- Register value preservation

## Next Steps
- Consider implementing I-type operations
- Add more complex test cases
- Implement branch operations

### CODE REVIEW

File: app/riscv.ld [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  .text : {
+    *(.text.init)             /* Initialization code (if any) */
+    *(.text)                  /* All code sections */
+  } > RAM
+
+  .rodata : {
+    *(.rodata)                /* Read-only data */
+  } > RAM
+
+  .data : {
```
Comment (amichai-bd) [inline app/riscv.ld:19]: It's strange that you don't have explicit offset for dmem (dmem is everything that is it "text")

File: app/riscv.ld [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/* Minimal RISC-V linker script */
+OUTPUT_ARCH("riscv")         /* Specify the target architecture */
+ENTRY(_start)                 /* Set the entry point to _start */
+
+MEMORY {
+  RAM (rwx) : ORIGIN = 0x00000000, LENGTH = 4K /* Define a 4KB RAM starting at address 0 */
```
Comment (amichai-bd) [inline app/riscv.ld:6]: I would expect a dedicated d.em region

File: app/test.S [hunk 3]
```diff
@@ -0,0 +1,40 @@
+.section .text
+.global main
+
+main:
```
Comment (amichai-bd) [inline app/test.S:4]: I would expect you to have a test that includes also load and store commands

File: source/decoder/decoder.sv [hunk 4]
```diff
@@ -5,25 +5,39 @@ module decoder
 import cpu_pkg::*;
 (
     input logic [31:0] instruction,  // Raw 32b instruction from memory
+    input logic branch_cond_out,
```
Comment (amichai-bd) [inline source/decoder/decoder.sv:8]: Branch condition met is a better.

File: source/decoder/decoder.sv [hunk 5]
```diff
@@ -38,14 +52,143 @@ always_comb begin
...
     ctrl.mem_wr_en           = 1'b0;         // No memory write
     ctrl.sel_wb              = 1'b1;         // Select ALU for writeback
+    ctrl.sign_ext            = 1'b0;         // No sign extension
     
     // Set instruction-specific control signals
     case (instruction[6:0])
-        7'b0110011: begin // R-type (ADD)
+        7'b0110011: begin // R-type 
             ctrl.sel_alu_imm = 1'b0; // Use register for ALU input 2
+            case(instruction[14:12]) // funct3
```
Comment (amichai-bd) [inline source/decoder/decoder.sv:61]: I would add both the funct 7 and funct 3 and concatenation them in the "case".
Case ({funct7,funct3})

**Review by amichai-bd — APPROVED** [review]
Kick ass code
